### ISSUE TITLE
move each toolkit wrapper implementation into its own module

### ISSUE THREAD

**Issue by adalke** [openforcefield/openff-toolkit#988]
The toolkit wrappers are in openff/toolkit/utils/toolkits.py I find that file very large (nearly 6,000 lines!) and a bit hard to understand and navigate because there are implementations for RDKit, OpenEye, and AmberTools all in the same file.

Not only is file is large, functions have internal imports (to defer importing modules that might not exist), there are classmethod and staticmethod solely to get around the stick-everything-into-a-class approach, and I get confused sometimes looking at a function like "to_object()" which is defined three times; I have to pay close attention to which one I'm looking at.

**Describe the solution you'd like**

I propose splitting the functions for the three wrapper implementation each into their own modules, like "openeye_wrapper.py", "rdkit_wrapper.py", and "ambertools_wrapper.py". There will still be an OpenEyeToolkitWrapper, RDKitToolkitWrapper, and AmberToolsToolkitWrapper class. On instantiation they will import the corresponding module, and load the appropriate functions.

I have a prototype of how that might work, at https://github.com/openforcefield/openff-toolkit/tree/toolkit-impl . It has a new ModuleToolkitWrapper(ToolkitWrapper) class that does the appropriate forwarding.

There's a new `publish()` decorator, which indicates that a given module's functions are to be public via the ToolkitWrapper class. This helps make a stronger distinction between the private API/internals and the public API.

The test_toolkits.py tests all pass with this change. I had to modify a couple of cases:

1) there is no longer a _check_license() test for OpenEye via mocks as that requires dynamic testing, while the public API to OpenEyeToolkitWrapper only has does the check at instantiation time. The new code only does that check at initial import time

2) I had to instantiate a few RDKitWrapperToolkit instances because the tests assumed a staticmethod to access some internal test functions. I left the tests as RDKitWrapperToolkit()._method_call and used "publish" to make those calls be available. A better solution would be to unit test the underlying (private) functions directly, in rdkit_wrapper.py.

I believe this layout is easier to understand and use, but I can be wrong. Take a look for yourself. 

**Describe alternatives you've considered**

Could leave it as-in.

A few months ago I explored a two-module solution, where you would import the public-facing module, and the toolkit-specific code was in a private implementation module. My thought was that the "import rdkit", "import openeye" could be pushed into the implementation modules, which are only imported if the public module said it was okay. In practice this was clumsy and hard-to-understand.

The `toolkit-impl` branch uses another technique to avoid this two-step process. I import something like rdkit.Chem or openeye.oechem. If the import fails, I instead assign Chem, or oechem, etc. to a simple object that raises an informative ToolkitUnavailableException error for any attribute lookup.

 I also found that expressing the toolkit API through a module, rather than an class or instance, meant that I couldn't use inheritance to share code, or implement a `__repr__()`

**Comment by lilyminium** [comment]
I haven't had much time lately to look at this code (or your other replies to my issues, thank you!) but this has always been a question for me, yeah, why one couldn't simply define each class in a module, do the required checks in that module, and make the class available in toolkits.py with:

```
try:
	from .openeye_toolkit import OpenEyeToolkitWrapper
except LicenseError:
	pass
else:
	AVAILABLE_TOOLKITS.append(OpenEyeToolkitWrapper)
```

**Comment by adalke** [comment]
The specific example you showed wouldn't work because the from module import obj_name gets obj_name from `module.__dict__` with no way to interpose a license check. There's no equivalent for a module `__getattr__`. Though that can be implemented by interesting an instance into sys.modules, neither that nor the exception handling mechanism you proposed are common, and neither are they more useful than calling a function for the license check:
```
from . import openeye_toolkit 
if openeye_toolkit.is_licensed():
    AVAILABLE_TOOLKITS.append(openeye_toolkit.OpenEyeToolkitWrapper)
```

The code I prototyped maintains backwards compatibility, though incompletely. (It fails for code which calls a class's staticmethod without instantiating the class, which is done by the test code to call `_`-prefixed internal functions for unit testing. I believe that is outside the public-facing design of the current API.)

I  believe the existing 3-part module/class/instance system was designed so the following would be true: 1) subclasses can inherit shared docstrings and support routines, 2) the license check protocol was made part of the constructor so that "can I instantiate it?" and "can I use it?" become synonymous, 3) and therefore instance methods could import third-party modules with the assurance that the module would be there and licensed.

A consequence of this design was the all-or-on license check in the constructor meant that people couldn't use OpenEye-based OpenFF functionality unless they were all licensed, even if certain functions only needed, say, an OEChem license. See https://github.com/openforcefield/openff-toolkit/pull/850 . That fix allows OpenEyeToolkitWrapper() creation if any component is licensed.

It could be replaced with a singleton-based design, with an API change so that license checks and class instantiation are not identical.

The straight-forward implementation would break existing code.

If there is a desire to switch to a singleton solution, without breaking backwards compatibility, then I believe that's possible using `__new__`, like this:

```
class _RDKitToolkitWrapper(ModuleToolkitWrapper):
    _toolkit_name = "The RDKit"
    module_name = "openff.toolkit.utils.rdkit_wrapper"

rdkit_toolkit_wrapper = _RDKitToolkitWrapper()

class RDKitToolkitWrapper(_RDKitToolkitWrapper):
   def __new__(cls):
     import warnings
     warnings.log("Creating RDKitToolkitWrapper instances is deprecated. Please use rdkit_toolkit_wrapper instead.")
     rdkit_toolkit_wrapper.license_check()
     return rdkit_toolkit_wrapper
```

This would mean subclass tests like `isinstance( rdkit_toolkit_wrapper, RDKitToolkitWrapper)` would still work.

However, it should only be part of a long-term strategy to migrate away from the existing API to a singleton, and I am not qualified to decide if that is appropriate for the project.

**Comment by lilyminium** [comment]
> The specific example you showed wouldn't work because the from module import obj_name gets obj_name from module.__dict__ with no way to interpose a license check.

Loading the module executes the code so you could just put the license check in the module. e.g., in polymetrizer I sprinkle "from oefuncs import function_name" knowing full well it doesn't exist and will fall down to rdfuncs. 

https://github.com/lilyminium/polymetrizer/blob/5cb1511b6f5a9c4df43dc7dbe42076cc02420b1d/polymetrizer/polymetrizer.py#L13-L16

**Comment by adalke** [comment]
In general, people expect that "import blah" is either successful or raises a ModuleNotFoundError or more generally an ImportError.

If I use polymetrizer but don't know if all of its dependencies are there, I can do:

```
try:
  import polymetrizer
except ImportError:
  polymetrizer = None
...
if polymetrizer is None:
   print("I’m sorry Dave. I’m afraid I can’t do that.")
else:
  polymetrizer.open_pod_bay_doors()
```

Or, I can fail only if I try to use polymetrizer functionality:

```
import polymetrizer
polymetrizer.open_pod_bay_doors()  # -> raises license exception
```

In principle you could document that the importation raises some other exception, and have people do:
```
try:
  import polymetrizer
except (ImportError, LicenseError):
  polymetrizer = None
```
but people aren't going to look for it, and it's more difficult to document or test. 

After all, OEChemToolkitWrapper might be "licensed" because a valid OEChem exists, but your code uses oequapac , which isn't licensed, so you or your API users likely have to handle license checks on a more granular level anyway than an all-or-nothing module import.

**Comment by SimonBoothroyd** [comment]
@adalke see also https://github.com/openforcefield/openff-toolkit/issues/834 and https://github.com/openforcefield/openff-toolkit/issues/966#issuecomment-852909992

**Comment by lilyminium** [comment]
@adalke That's right, sorry for not clarifying -- in my ideal implementation, [LicenseError would be a subclass of ModuleNotFoundError](https://github.com/openforcefield/openff-toolkit/issues/771#issuecomment-853507688). 

> python -O

That's right, yeah. Referring instead to your code:
```
try:
    from rdkit import Chem
except (ModuleNotFoundError, ImportError):
    Chemf = RDKitModuleUnavailable()
    _is_installed = False
```

you could instead trigger it directly:

```python
RDKitModuleUnavailable()
```

> There will still be an OpenEyeToolkitWrapper, RDKitToolkitWrapper, and AmberToolsToolkitWrapper class. On instantiation they will import the corresponding module, and load the appropriate functions.

May I ask why have this step, instead of just putting the OpenEyeToolkitWrapper etc. in a separate file? Just because it *is* a little strange for this to be module-based for me -- many different objects with the same API does feel like it suits the class model.

**Comment by adalke** [comment]
"Different objects with the same API" also fits the module model. In Python, compare the `[math](https://docs.python.org/3/library/math.html)` and `[cmath](https://docs.python.org/3/library/cmath.html)` modules. In cheminformatics toolkit wrapper space, see [cinfony](http://cinfony.github.io/).

The class model comes with several additional concepts:

1) class-vs-instance: OpenFF effectively only uses the toolkit wrappers as singletons. While the implementation does create different instances, they all act the same, and could be replaced by the same object. The `__init__` is used as hook to test if the toolkit is available and (if licensed) at least one license is found. However a) that's an unusual use of a constructor, b) it's too coarse-grained, as we see from OEChem component licensing, and c) other methods work equally well.

2) inheritance: Nothing in OpenFF needs inheritance and I don't see any case where it's useful to derive from (say) OpenEyeToolkitWrapper. Rather, I think the "toolbox" concept I proposed at https://github.com/openforcefield/openff-toolkit/issues/966#issuecomment-863352782 is more flexible, as it allows mixins in a way that inheritance doesn't.

3) mixins: classes can be used for mixins, as in `BuiltInToolkitWrapper._check_n_conformers`. However, in this case the class mechanism is used only as a namespace. In the prototype, I replaced a `self._check_n_conformers` with `BuiltInToolkitWrapper._check_n_conformers`.  Modules can handle that equally well. (IMO, a bit better as there's no need for the `@staticmethod` decorator.)

4) unit testing: I think that classes are worse for OpenFF's needs than module functions. Classes either require you instantiate the object (which, as I pointed out, are singleton), or use static or class method annotations. Consider this test from `test_toolkits.py`:

```
    def test_elf_is_problematic_conformer_acid(
        self, formic_acid_molecule, formic_acid_conformers
    ):
        problematic, reason = RDKitToolkitWrapper._elf_is_problematic_conformer(
            formic_acid_molecule, formic_acid_conformers["cis"]
        )
```

This works because `_elf_is_problematic_conformer` is a `@classmethod`. It uses the class to call `cls._find_smarts_matches`, which is a `@staticmethod`. It's a complex dance between instance methods, class method, and static methods. 

My prototype failed in this case, because its methods don't exist until instantiation. In my prototype, I made this example work by always instantiating `RDKitToolkitWrapper`:

```
    def test_elf_is_problematic_conformer_acid(
        self, formic_acid_molecule, formic_acid_conformers
    ):
        problematic, reason = RDKitToolkitWrapper()._elf_is_problematic_conformer(
            formic_acid_molecule, formic_acid_conformers["cis"]
        )
```
If this proposal is accepted, then the test should be refactored to:
```
    def test_elf_is_problematic_conformer_acid(
        self, formic_acid_molecule, formic_acid_conformers
    ):
        problematic, reason = rdkit_wrapper._elf_is_problematic_conformer(
            formic_acid_molecule, formic_acid_conformers["cis"]
        )
```
which, in essence, treats everything as class+static methods. Which is fine. And it's a clear pointer that it's calling into the internals of the module, rather than going through the toolkit API. (NOTE: I think these RDKit-specific tests should be in their own test module for that reason.)

For these reasons, I don't see that the class model is clearly better than the module model + class wrapper, while I think the latter has some advantages.

**Comment by adalke** [comment]
> you could instead trigger it directly:
> 
> ```python
> RDKitModuleUnavailable()
> ```

That's certainly one of several approaches. The next step is to figure out which other factors should influence the choice of the final decision.

For example, my approach means rdkit_wrapper can be imported even if RDKit isn't installed. My preference would be to let the "from rdkit import Chem" ModuleNotFoundError raise right to the top, with no modification. Then toolkits.py would catch that exception and configure things accordingly, also with possible warning messages.

However, the existing RDKitToolkitWrapper class lets people see what methods are available, even if RDKit cannot be loaded, eg, by dir(RDKitToolkitWrapper). A suitable IDE, including '?' in Jupyter, will show the docstrings. These can't happen if there's an import error, or a raised RDKitModuleUnavailable error.

Is this *important*? I can't say of it is for this project. My view is that it's a nice-to-have, but not critical. Still, the feedback I've heard is that despite comments in the code like "This API is experimental and subject to change", there is a strong desire not to break compatibility without a strong reason. And this doesn't seem strong enough.

What does an RDKitModuleUnavailable() exception add to the default exception? The two things I can see are 1) fine-grained exception handling, and 2) a pointer to how to install via conda.

My experience is that fine-grained exceptions fall into the classification problem Borges satirized in 'Celestial Emporium of Benevolent Knowledge' - they are so rarely needed that it's almost never worth the effort.

Consider 1)  An RDKitModuleUnavailable means the importer who knows about OpenFF's new subclass could catch the exception and not parse the "No module named 'rdkit'" message that would normally occur, if that were important. But that saves one line of code. On the other hand, an importer who didn't know about OpenFF's new subclass, and expected an ImportError, would be unable to parse the new error message because it's in a different format.

Consider 2) People use RDKit from sources other than conda. For example, there's a recent pip-installable package, and I compile it from scratch. And that message isn't shown if the ImportError is caught - but shouldn't it be for that case? I think the error message should be sent to stderr (or a logger with a default setting to stderr), and only shown once in the lifetime of the program. Raising a RDKitModuleUnavailable() with installation instructions in it is either going to be shown once, with the exception reaches the top-level of the program and causes it to exit (the current design purpose), or be swallowed up every single time.

As you can infer, I'm not a fan of the current exception system. I don't see what something like ToolkitUnavailableException adds, and I think having it leads to oddities, like how deregister_toolkit raises a ToolkitUnavailableException if the toolkit is available but had already been deregistered. (It's quite logical if you interpret it to mean "toolkit unavailable in the current registry", but that's different from intended "toolkit unavailable to the Python importer").

**Comment by lilyminium** [comment]
This has diverted into exception-discussion a little, which maybe belongs more in #771. In that issue I think @mattwthompson and I agree that most exceptions should be subclasses of other exceptions, so `RDKitModuleUnavailable` should be some version of `ImportError` too. I do think that having custom install instructions is extra unnecessary work in the age of Google.

Thanks for the chatting about modules and classes, I was not creative enough in looking past ToolkitWrappers in the first place. For a puddle of functions then that does intuitively feel like a module. Inheritance is less of an issue because one could just import all functions from openeye_module.py and mimic inheritance that way. One thing I *do* like about classes is the amount of futzing you can do with metaclasses under the hood and without affecting the API. For example, you could automatically register new modules / toolkitwrappers upon class creation. That's the only good use I can think of for a metaclass currently so it is probably overengineering for now.

One nice but nonessential thing about toolbox is the possibility of setting up a registry of toolkits that have each method, e.g. current_toolbox.available["find_smarts_matches"] = [rdkit, openeye]

Another nice but nonessential thing would be hooking the modules to strings, it would be really great to `Molecule.assign_partial_charges(toolkit="rdkit")` instead of remembering whichever very long path to import the relevant wrapper or toolkit or toolbox from.

**Comment by mattwthompson** [comment]
I'm +1 on moving each class into its own file.

I'm +1 on specific exceptions as catching specific exceptions (without needing to inspect STDRR) is hugely for downstream users. I agree we should continue that discussion elsewhere.

I've lost track of the other discrete topics(s) here, and am inclined to suggest we break changes up into discrete pieces wherever possible. It's fine for discussions to branch out and cover many points, but that's not a workable way to implement patches.

**Comment by adalke** [comment]
> One nice but nonessential thing about toolbox is the possibility of setting up a registry of toolkits that have each method, e.g. current_toolbox.available["find_smarts_matches"] = [rdkit, openeye]

What is the use case? That is, what makes it nice?

I've updated my prototype since the gist. The current version allows `get_super()` on the methods, which gets the next handler in the resolution chain.

```
>>> toolbox_prototype.current_toolbox.find_smarts_matches.__module__
'openff.toolkit.utils.openeye_wrapper'
>>> toolbox_prototype.current_toolbox.find_smarts_matches.get_super().__module__
'openff.toolkit.utils.rdkit_wrapper'
>>> toolbox_prototype.current_toolbox.find_smarts_matches.get_super().get_super()
```

Thus allows functionality like:
```
>>> implemented_by(toolbox_prototype.current_toolbox.find_smarts_matches)
>>> def implemented_by(func):
...   names = []
...   while func is not None:
...     names.append(f"{func.__module__}:{func.__qualname__}")
...     func = func.get_super()
...   return names
...
>>> implemented_by(toolbox_prototype.current_toolbox.find_smarts_matches)
['openff.toolkit.utils.openeye_wrapper:find_smarts_matches', 'openff.toolkit.utils.rdkit_wrapper:find_smarts_matches']
```
Note that in the current system, the same function may occur several times in the resolution chain. There are ways to de-duplicate this, if important.

> Another nice but nonessential thing would be hooking the modules to strings, it would be really great to `Molecule.assign_partial_charges(toolkit="rdkit")` instead of remembering whichever very long path to import the relevant wrapper or toolkit or toolbox from.

The toolbox approach would make this more like:

```
with toolbox("rdkit"):
    mol.assign_partial_charges()
```

**Comment by lilyminium** [comment]
> What is the use case? That is, what makes it nice?

Right now to find which toolkits implement a particular method I ctrl+f inside the toolkits.py file, which is not the best solution. This became relevant when I was trying to work out which toolkits implemented smarts matching and which implemented generating conformers.

**Comment by adalke** [comment]
> This became relevant when I was trying to work out which toolkits implemented smarts matching and which implemented generating conformers.

An advantage to having a central location is that  everything is, well, centralized. A hypothetical  system might look like:

- `current_toolbox.<tab>` to get a list of the available methods (available)
- `current_toolbox.generate_conformers.get_implementations()` - (hypothetical) the actual implementation functions
- use any one of various methods to get information about the implementation functions

BTW, I think "toolkit" is too specific. What if you added your own conformation generation tool and wanted to use it instead? Is that a toolkit? Perhaps yes. But the point of my distinction between "toolkit" and "toolbox" is to separate a "provider of functions" vs. a "selection of functions", and the current toolkit-oriented system makes it tricky to swap selections in and out.

**Comment by SimonBoothroyd** [comment]
Thanks for putting this together @adalke - I think the idea is definitely interesting but I'm not sure I'm 100% sold on it just yet.

> Right now to find which toolkits implement a particular method I ctrl+f inside the toolkits.py file

+1 to this point - this is mainly why I think the way toolkits are implemented could probably be changed (and likely in a way which is not API breaking in most cases). 

I think the approach proposed here does not really resolve this however, as you'd still need to know that 'SMARTS matching support = I need to check if `find_smarts_matches` is implemented / exists' (which gets harder to keep track the more cases you want to check for, is harder for static linters to validate, and provides no guarantee about the function signature, etc.) rather than - is there are toolkit wrapper which implements a well defined 'SMARTS matching provider' mix-in base class.

Ideally the toolkit would outline the key 'functionality' it needs to provide (i.e. charge assignment + WBO comp, conformer gen, SMARTS matching, aromaticity models, stereo models) etc. and just provide a very well defined interface for that functionality based on abstract class inheritance + the `abc` module.

You could still have a central 'toolbox / registry' access point which houses the individual wrapper / providers (rather than decorated functions), but instead it would have fixed, well defined, lintable methods for retrieving the `'current_registry.charge_providers'` or `'current_registry.conformer_providers'` and similar.

To an extend I think the 'get_super()` infrastructure is also kind of misleading because you aren't really allowing for inheritance here in a true sense, more a chain of fallback options. 

Storing methods on methods also seems a little odd and not particularly pythonic, but that's maybe an opinionated view point.

> current_toolbox.<tab>

I've not checked yet but dynamically created / populated objects tend not to interact too well with code completion tools in IDEs - I'd be interested to see how well PyCharm handles this design pattern.

> Perhaps yes. But the point of my distinction between "toolkit" and "toolbox" is to separate a "provider of functions" vs. a "selection of functions", and the current toolkit-oriented system makes it tricky to swap selections in and out.

+1 to this - I'm not 100% convinced though that have a collection of 'any function you want to add' is necessarily the cleanest way to go. My +1 is still - a toolkit registry / toolbox should state which 'providers' (toolkit registries currently) are available for each core functionality needed by the toolkit.

A big +1 for any solution that lets me state using a config file which order of preference to use for each feature set.

**Comment by lilyminium** [comment]
> What if you added your own conformation generation tool and wanted to use it instead? Is that a toolkit?

That could be one argument for the structure of metaclasses.

```python
AVAILABLE_TOOLKITS = {}
AVAILABLE_TOOLKIT_METHODS = defaultdict(list)

class ToolkitMeta(type):
	def __init__(self, name, bases, clsdict):
		AVAILABLE_TOOLKITS[name] = self  # or self() for singleton?
		for method_name, method in clsdict.items():
			AVAILABLE_TOOLKIT_METHODS[method_name].insert(0, method)  # probably want latest definitions first

class ToolkitBase(metaclass=ToolkitMeta):
	pass

class MyCustomConformerGenerationToolkit(ToolkitBase):

	def generate_conformers(self, *args, **kwargs):
		print("or something along these lines")
```

This is just spitballing for fun, not really a serious suggestion.

**Comment by lilyminium** [comment]
Apologies for being instrumental in derailing this issue. The topics/nuances discussed here seem to be:
- Split up toolkits.py (+1 by me, @SimonBoothroyd, @mattwthompson) (#834)
- Module vs. classes for providing toolkit functions (no consensus)
  - Inheritance
  - abstract base class interfaces
  - singleton vs normal classes
- Needing a solution for specifying which backend to use per function (#966)
- Code completion tools in IDEs (as opposed to interpreters)
- Needing some way to impose a standard API
  - [Optional] Needing some way for users to create their own toolkits/functions/whatever to splice in
  - [Optional] Needing a way to figure out what current implementations exist per function
  - [Optional] Some way to specify toolkit/backend by string, please
- Exceptions => #771


----
Side opinions:

> abstract base class interfaces

The current toolkit set up has different backends with different functionalities, though. If you propose the kind of ABC that raises an error on non-implementation, how should AmberTools handle conformer generation?

> Needing some way to impose a standard API

What about a Toolkitboxthing that has the public function signature and just passes those public args to whichever backend is selected? It's up to the backend implementation how to make use of the given arguments.

**Comment by SimonBoothroyd** [comment]
> If you propose the kind of ABC that raises an error on non-implementation, how should AmberTools handle conformer generation?

You can use [mix-ins](https://github.com/openforcefield/openff-toolkit/issues/966#issuecomment-852909992), so essentially the AT wapper would only inherit from the charge provider base class and not inherit from the conformer base class. The OE wrapper would inherit from both etc.

**Comment by lilyminium** [comment]
> You can use mix-ins

I don't know if that would be an ideal solution. You basically have to manually trigger your own error by remembering to inherit from that class yourself. What stops something inheriting from only the charge provider base but still having the conformer generation method?

**Comment by SimonBoothroyd** [comment]
> You basically have to manually trigger your own error by remembering to inherit from that class yourself.

I'm not sure I understand your point, can you clarify?

> What stops something inheriting from only the charge provider base but still having the conformer generation method?

Nothing. The design pattern I'm suggesting is similar to interfaces in Java and C#, in so much as your base class simply defines the contract for exactly what a feature provider must specify - i.e. the exact functions and their non-optional signature. Subclasses can implement whatever else they want, all the interface base class is doing is explicitly stating that subclasses can safely be used for this purpose with no extra guarantees.

To a large extent this is mostly sugar for static linters and code completion IDEs (e.g. you can use the interface base class as the type hint for the return value of `current_registry.charge_providers`) - python will always be a duck typed language but still it's nice to try and be as precise as possible re. what is supported / required.

**Comment by adalke** [comment]
> I think the approach proposed here does not really resolve this however, as you'd still need to know that 'SMARTS matching support = I need to check if `find_smarts_matches` is implemented / exists' (which gets harder to keep track the more cases you want to check for, is harder for static linters to validate, and provides no guarantee about the function signature, etc.) rather than - is there are toolkit wrapper which implements a well defined 'SMARTS matching provider' mix-in base class.

mypy handles [structural subtyping](https://mypy.readthedocs.io/en/stable/protocols.html), which handles most of the static linting requirements I think you'll have, and I think also more cleanly than having all toolkits derive from all combinations of interfaces they implement. Think that one bit of functionality might need {A, B, C}, another {A, B, D}, an another {A, C, D}, and another just {A}.

"guarantee about the function signature" is not hard to enforce. I can think of several approaches, depending on your specific requirements. (Eg, verify every time a function with certain names is added to a toolbox, or verify when creating a function then adding annotations that it's safe.)

BTW, what exactly are your requirements? inspect.getfullargspec() must match? Or can `*args, **kwargs` be allowed? Can an implementation add additional parameters? 

For example:
```
OpenEyeToolkitWrapper:    def apply_elf_conformer_selection(
        self,
        molecule: "Molecule",
        percentage: float = 2.0,
        limit: int = 10,
    ):
RDKitToolkitWrapper:      def apply_elf_conformer_selection(
        self,
        molecule: "Molecule",
        percentage: float = 2.0,
        limit: int = 10,
        rms_tolerance: unit.Quantity = 0.05 * unit.angstrom,
    ):
```
Is that an expression of a base implementation, with a derived extension for `rms_tolerance`? Or should this not be allowed? Or should OpenEyeToolkitWrapper accept but ignore that field? Or might each method have different requirements?
 
> To an extend I think the 'get_super()` infrastructure is also kind of misleading because you aren't really allowing for inheritance here in a true sense, more a chain of fallback options.

Okay, "get_next_implementation()". 

> Storing methods on methods also seems a little odd and not particularly pythonic, but that's maybe an opinionated view point.

I have no problems with using a callable instance instead, that is, something with a `__call__`. A concrete prototype exists in part to show what  can be done, and identify rough edges. 

Function-with-methods is a hack. My original version had only the function call, then I realized that having `get_super()` attached to it made the debug example much cleaner, and made it possible for people to inspect the entire chain. Adding the existing bound method to the function took only a few lines.


> I've not checked yet but dynamically created / populated objects tend not to interact too well with code completion tools in IDEs - I'd be interested to see how well PyCharm handles this design pattern.

It should not be a problem. I use [`__dir__`](https://docs.python.org/3/reference/datamodel.html?highlight=__dir__#object.__dir__) which is the standard way to implement this feature. See for example the implementation of `inspect.getmembers()`, which early on does `names = dir(object)`. Also, tab completion in the Python terminal is implemented by rlcompleter.py, which uses `words = set(dir(thisobject))`

Let me know if it is a problem - I don't have PyCharm.

> A big +1 for any solution that lets me state using a config file which order of preference to use for each feature set.

Like this?
```
# my_config.py -- or in your own usercustomize, or I can
# think of a few other possibilities.
feature_set1 = toolbox("rdkit", "openeye", from_smiles = my_func)
feature_set2 = toolbox("openeye", keep_looking_ = False)  # was 'inherit_'
feature_set3 = toolbox("rdkit", generate_conformers = super_duper_conf_generation)
  ...
from my_config import *
import random
feature_set = random.choice([feature_set1, feature_set2, feature_set3]")
with feature_set:
    mol = current_toolkit.from_smiles("C#N")
    mol.generate_conformers()
```

Perhaps "feature_set" is a better term than "toolbox"?

**Comment by j-wags** [comment]
Wow! A big discussion here across a lot of topics. This touches so many things that I'm not sure how to break this down. Thanks @lilyminium for trying to discretize the discussion, that's a big help.


>     Split up toolkits.py (+1 by me, @SimonBoothroyd, @mattwthompson) (#834)

+1 from me as well. I took a look at this PR, it's so large that the file diff is meaningless. But the tests pass and a quick glance at the docs looked good. I'll look at this again with fresh eyes and will probably approve!

>    Module vs. classes for providing toolkit functions (no consensus)
>        Inheritance
>        abstract base class interfaces
>        singleton vs normal classes

I'm still thinking on these. 

I found the discussions here about what each implementation would ENABLE to be most helpful. So I'm going to extend @lilyminium's other points with a sort of (unsorted) scorecard as I read over this discussion a few more times and check out the proposed implementations. My rough priorities are

* Doesn't break the API where there's not a clear benefit
* Be readable, debuggable, and extensible by contributors
* Compatible with code completion tools in IDEs and interpreters
* Enforcing and exposing method APIs
  * Impose a standard API
  * Allow users to create their own toolkits/functions/whatever to splice in
  * Allow users to figure out what current implementations exist per function
  * Allow toolkit methods that add additional keywords past the standard API
  * Convenient ways to "publish" methods on other classes (like Molecule)
* Handling multiple cheminformatics toolkits
  * Allow users to specify which backend to use per function (#966) 
  * Let users see in advance which methods will be called, in which order
  * Have a "just does the right thing for most users" default registry
  * Let users select order for method calls, potentially using a config file, also potentially using a context manager
  * Let users query provenance from the RESULT of toolkit calls (even complex nested ones!)
  * Some way to specify toolkit/backend by string
  * Graceful partial-license handling for eg. OpenEye
* Resolve the weird "ambertools needs rdkit for conf gen" situation (where either a toolkit needs to hard-code another toolkit, or it needs to refer back to the current context to figure out which other toolkit should do something)
* Simplify the "special handling of partial charges specifically when going to/from SDF" thing
* Not block implementation of caching

**Comment by adalke** [comment]
> > ```
> > Split up toolkits.py (+1 by me, @SimonBoothroyd, @mattwthompson) (#834)
> > ```
> 
> +1 from me as well. I took a look at this PR, it's so large that the file diff is meaningless. 

@j-wags - there are two PRs. I think you should look at https://github.com/openforcefield/openff-toolkit/issues/834 with my PR at https://github.com/openforcefield/openff-toolkit/pull/993 .

All that one does is split the current monolith into multiple submodules, with imports to keep the existing API unchanged.

**Comment by lilyminium** [comment]
> I'm not sure I understand your point, can you clarify?

> all the interface base class is doing is explicitly stating that subclasses can safely be used for this purpose with no extra guarantees.

> To a large extent this is mostly sugar for static linters and code completion IDEs (e.g. you can use the interface base class as the type hint for the return value of current_registry.charge_providers)


I may be misunderstanding you, but I think you were proposing that subclass toolkitwrappers inherit from a parent mixin class with an @abc.abstractmethod so that errors are raised unless the subclass implements said method. However, unless I actively choose to subclass `AmberTools(ChargeMethodsMixin)`, it won't raise errors for the charge-related methods. At that point since I'm aware enough of the `ChargeMethodsMixin` to know that I should subclass it, I didn't see much difference in skipping the middleman and just implementing the methods without inheriting from the mixin (assuming it doesn't come with other niceties like standard methods for every toolkit). However, it would help IDEs for return values where the value is some kind of subclass, as you pointed out.

**Comment by lilyminium** [comment]
Was this closed by #993?

**Comment by j-wags** [comment]
I think so. There's a lot of really valuable discussion here on topics beyond the title, but it'll still be accessible if the issue is closed.
